core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -3104,23 +3104,103 @@ public static int position(ByteString seek, ByteString s) {
 
   /** SQL {@code POSITION(seek IN string FROM integer)} function. */
   public static int position(String seek, String s, int from) {
-    final int from0 = from - 1; // 0-based
-    if (from0 > s.length() || from0 < 0) {
+    if (from == 0) {
+      throw new IllegalArgumentException("From position cannot be zero");
+    }
+    // Case when from is positive
+    if (from > 0) {
+      final int from0 = from - 1; // 0-based
+      if (from0 > s.length() || from0 < 0) {
+        return 0;
+      }
+
+      return s.indexOf(seek, from0) + 1;
+    }
+    // Case when from is negative
+    final int rightIndex = from + s.length(); // negative position to positive index
+    if (rightIndex <= 0) {
       return 0;
     }
-
-    return s.indexOf(seek, from0) + 1;
+    return s.substring(0, rightIndex+1).lastIndexOf(seek) + 1;
   }
 
   /** SQL {@code POSITION(seek IN string FROM integer)} function for byte
    * strings. */
   public static int position(ByteString seek, ByteString s, int from) {
-    final int from0 = from - 1;
-    if (from0 > s.length() || from0 < 0) {
+    if (from == 0) {
+      throw new IllegalArgumentException("From position cannot be zero");
+    }
+    // Case when from is positive
+    if (from > 0) {
+      final int from0 = from - 1; // 0-based
+      if (from0 > s.length() || from0 < 0) {
+        return 0;
+      }
+      return s.indexOf(seek, from0) + 1;
+    }
+    // Case when from is negative
+    final int rightIndex = from + s.length();
+    if (rightIndex <= 0) {
       return 0;
     }
+    int lastIndex = 0;
+    while (lastIndex < rightIndex) {
+      int indexOf = s.substring(lastIndex, rightIndex + 1).indexOf(seek) + 1;
+      if (indexOf == 0) {
+        break;
+      }
+      lastIndex += indexOf;
+    }
+    return lastIndex;
+  }
+
+  /** SQL {@code POSITION(seek, string, from, occurrence)} function. */
+  public static int position(String seek, String s, int from, int occurrence) {
+    if (occurrence == 0) {
+      throw new IllegalArgumentException("Occurrence cannot be zero");
+    }
+    for (int i = 0; i < occurrence; i++){
+      if (from > 0){
+        from = position(seek, s, from + (i == 0 ? 0 : 1));
+        if (from == 0) {
+          return 0;
+        }
+      } else {
+        from = position(seek, s, from);
+        if (from == 0) {
+          return 0;
+        }
+        from -= (s.length() + 2);
```

Comment:
Don't know if this could be helpful but it seems like there's a lot of checking to determine if FROM is a valid value or not, could a helper be useful? This logic is crazy so I don't know if it would help that much or not, just a thought.

Reply:
yeah I think there's definitely some simplifying that could be done, especially with all the +1s -1s and the like flying around. I'll look into it